dashboard/app/models/script.rb

```diff
@@ -1125,55 +1125,63 @@ def fix_lesson_positions
+          UnitGroupUnit.create!(unit_group: destination_unit_group, script: copied_unit, position: destination_unit_group.default_units.length + 1)
+          copied_unit.reload
+        else
+          copied_unit.is_course = true
+          raise "Must supply version year if new unit will be a standalone unit" unless version_year
+          copied_unit.version_year = version_year
+          raise "Must supply family name if new unit will be a standalone unit" unless family_name
+          copied_unit.family_name = family_name
+          CourseOffering.add_course_offering(copied_unit)
+        end
 
-      if destination_unit_group
-        raise 'Destination unit group must be in a course version' if destination_unit_group.course_version.nil?
-        UnitGroupUnit.create!(unit_group: destination_unit_group, script: copied_unit, position: destination_unit_group.default_units.length + 1)
-        copied_unit.reload
-      else
-        copied_unit.is_course = true
-        raise "Must supply version year if new unit will be a standalone unit" unless version_year
-        copied_unit.version_year = version_year
-        raise "Must supply family name if new unit will be a standalone unit" unless family_name
-        copied_unit.family_name = family_name
-        CourseOffering.add_course_offering(copied_unit)
-      end
+        lesson_groups.each do |original_lesson_group|
+          original_lesson_group.copy_to_unit(copied_unit, new_level_suffix)
+        end
 
-      lesson_groups.each do |original_lesson_group|
-        original_lesson_group.copy_to_unit(copied_unit, new_level_suffix)
-      end
+        course_version = copied_unit.get_course_version
+        copied_unit.resources = resources.map {|r| r.copy_to_course_version(course_version)}
+        copied_unit.student_resources = student_resources.map {|r| r.copy_to_course_version(course_version)}
 
-      course_version = copied_unit.get_course_version
-      copied_unit.resources = resources.map {|r| r.copy_to_course_version(course_version)}
-      copied_unit.student_resources = student_resources.map {|r| r.copy_to_course_version(course_version)}
+        # Make sure we don't modify any files in unit tests.
+        if Rails.application.config.levelbuilder_mode
+          copy_and_write_i18n(new_name, course_version)
+          copied_unit.write_script_json
+          destination_unit_group.write_serialization if destination_unit_group
+        end
 
-      # Make sure we don't modify any files in unit tests.
-      if Rails.application.config.levelbuilder_mode
-        copy_and_write_i18n(new_name, course_version)
-        copied_unit.write_script_json
-        destination_unit_group.write_serialization if destination_unit_group
+        copied_unit
       end
-
-      copied_unit
+    rescue => e
+      filepath_to_delete = Script.script_json_filepath(new_name)
+      File.delete(filepath_to_delete) if File.exist?(filepath_to_delete)
+      puts "Error: #{e.message}"
```

Comment:
what about level files -- could those persist (and get scooped) when the script copy fails?

Reply:
Those persist. Should we go ahead and clean those up? That's not what the ticket detailed but would be happy to do it, or create another ticket for it

Reply:
what do you think @bethanyaconnor ? I'm fine with merging this fix independently, but now would be a good time to figure out if we should also clean up the level files before moving on.

Reply:
Ideally, the files that are created should represent what was saved in the database so we should probably clean them up if they're not represented in the database. On the other hand, they don't cause as much harm as the script_json files being left around as the level files are created if cloning was successful on that particular level (it's unlikely for there to be any downstream failures).

I agree we should be decide if we want to clean up the level files but I don't necessarily think it should be part of this PR. Personally, I'm struggling with exactly _how_ we would clean up the level files, with this implementation or with a different one.

Reply:
Shall I create a ticket for it? Unless we see an easy way to do it now

Reply:
> I agree we should be decide if we want to clean up the level files but I don't necessarily think it should be part of this PR. Personally, I'm struggling with exactly how we would clean up the level files, with this implementation or with a different one.

I would imagine that "cleaning up" level files would actually look more like not writing them in the first place. perhaps something like this:
1. disable writing of level files during clone. ideas:
  * a. use a flag like `skip_name_format_validation`, which we use on Script to skip a validation. could similarly use this to skip the `after_save :write_custom_level_file`
  * b. take away `after_save :write_custom_level_file`, and instead explicitly call write_custom_level_file at the times that we want it to happen. this could be a bit messy and might lead to bugs, but would probably work in the end
2. do the deep script clone in a DB transaction
3. `script.levels.each(&:save!)` or `script.levels.each(&:write_custom_level_file)` or something else which takes care of also writing dsl-defined level files

having said that... even the easiest potential option (1a) seems like a fair bit of work and I'm fine with punting on it for now per your recommendation Bethany.

Reply:
> Shall I create a ticket for it? Unless we see an easy way to do it now

I am going to say yes please... because cloning is a tricky problem and it would be good to have access to any past ideas we've had for narrowing down the problem.

Reply:
Sounds good––see https://codedotorg.atlassian.net/browse/PLAT-1572